### DISCUSSION TITLE
Tables, Transforms and Pipelines

### DISCUSSION THREAD

**Discussion by snth** [PRQL/prql#1323] [General]
Recent developments on #523 reminded me that I've long wanted to have a discussion around clarifying terms and concepts around *tables*, *transforms* and *pipelines* and how there relate to CTEs and (table valued) functions.

I have developed a framework for myself around this and will be using the terms defined as follows (which doesn't correspond 100% to how we have been using these terms in discussions so far, particularly the term *pipeline*):
  * **table**: aka **DataFrame**, a *set* of tuples/rows
  * **transform**: transforms take a *table* as input and produce a new *table* as output
  * **pipeline**: a *sequence* of *pipelines* and *transforms* which produce a *table*

Transforms are made up of "atomic" transforms which correspond well to the transforms listed in section 3 of the book:
![image](https://user-images.githubusercontent.com/652285/208931894-8efaa646-0512-48b4-9ead-f2d774eceb67.png)

They mostly have a signature of `transform: table -> table` with one notable exception, namely `from` which is actually a pipeline (the atomic pipeline) and has signature `from: null -> table`.

So in my terminology we then have the following:
  * **table**: a struct type, with optional restrictions of columns that have to exist with given types, e.g.
    `employees<table[employee_id:int, name:str]>`
    `salaries<table[employee_id:int, salary:float]>`
  * **transform**: a function type with signature `<table -> table>`, e.gi
    `take<int table -> table>`
  * **pipeline**: a function type with signature `<null -> table>`, e.g.
    `from<str null -> table>` (We could argue about the null here. I initially didn't include the str in my signature. I just want to make it explicit that it doesn't act on a previous pipeline.)

---
Now one of the things that sets PRQL apart is the composability its transforms, with transforms being composed simply with `\n` or `|`. As of recent PR #1195, we can also put "pipelines" in lots of places which further increases PRQL's capabilities here. 

Being able to produce custom (and reusable) transforms in PRQL could really be one of the key features for the language. However in practice I think we're not quite there yet, and in my analysis this comes down to how "namespaces" are handled in SQL. I will try to illustrate this with some examples. I'll largely make up some syntax.

```elm
pipeline top_5_salaries = (
  from s=salaries
  sort [-s.salary]
  take 5
)
```
It would be great to be able to generalise this, so let's say we could parameterise the pipeline definition in which case it actually becomes a transform
```elm
transform take_top_salaries n<int> s<table[salary:float]> = (
  sort [-s.salary]
  take n
)
```

This looks reasonable and we might expect the following to work:
```elm
from salaries
take_top_salaries 5
```
Because we know the previous pipeline always gets passed to the transform as the last parameter, we should be able to make the identification `s=salaries`.

However what happens when we pass a pipeline with multiple source tables, e.g.
```elm
from employees
join salaries [==employee_id]
take_top_salaries 5
```
? It would not be clear which source table/namespace to take the `salary` column from. Maybe this is not an issue as long as the source pipeline fulfills the type restriction `table[salary:float]` of having a uniquely named column `salary` of type `float`.

Dataframe libraries like Pandas, which have a logical model closer to PRQL, get around this by forcing unique column names at join time, using automatic renaming schemes like `*lsuffix*` and `*rsuffix*` (see e.g. [pandas.DataFrame.join](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.join.html)). SQL is more lenient and let's all columns exist in `tablename.` prefixed namespaces.

---
This is as far as I've gotten in my explorations. I would welcome any thoughts, comments and hopefully discussion as it's something I've been trying to express for a while but only recently been able to write down in (hopefully) coherent form.

**Comment by snth** [comment]
As an aside, to me the whole thing smells of Monads but I'm too unfamiliar/rusty to really be able to express it properly so please read the following in that context.

My guesses are that:
  * **from** is the `unit`/`return` that takes a table name (object) and materialises it in a database Monad.
  * **PRQL transforms** are `functions`
  * The **compiled SQL** are monadic `actions` acting in the database monad.

Not sure if there is anything to be gained by making this identification.

**Reply by aljazerzen** [reply]
I cannot comment on that, I haven't delved too deep into Haskell / functional languages.

**Comment by aljazerzen** [comment]
A small correction on terminology, which I realized only recently: 

In context of databases, table usually means "persistently stored relation", while relation is the ordered set of tuples.

**Reply by snth** [reply]
Right, so a *view* would also be a *relation* and the difference to a *table* is that the table is persistently stored?

**Reply by aljazerzen** [reply]
Yes, exactly.

**Reply by max-sixty** [reply]
Ah OK great, I'll start using this too

**Comment by aljazerzen** [comment]
I think of the whole thing very similarly, except for this difference:

```diff
- pipeline: a sequence of pipelines and transforms which produce a table
+ pipeline: a sequence transforms which produce a relation
```


Relation (or table if I use your terminology) can then be either:
- a table, i.e. a source of data, or
- a pipeline


And the line between relation and transform is kinda blurry.
expression | type | name
----|-----|-----
`filter a > 3`  |                `relation -> relation`  |  transform
`filter`        |   `column<bool> relation -> relation`  |  also transform?
`from`          |   `ident -> relation`                  |  also transform?
`from my_table` |   `relation`                           |  relation (a single transfrom)
`from my_table \| filter a > 3` | `relation`            |  relation (two transfroms)

The difference in name may be only in whether you are asking "what is the value of this expression" as opposed to "what is the expression composed of"...

**Reply by snth** [reply]
I think it's important to clarify the terminology differences. I was using `table` where you use `relation`. I'm happy to adopt your terminology of `relation` for this discussion.

expression | type | name
-- | -- | --
filter a > 3 | relation -> relation | transform
filter | column<bool> relation -> relation | (parameterised) transform
from | ident -> relation | (parameterised) pipeline
from my_table | null -> relation | pipeline
from my_table \| filter a > 3 | null -> relation | pipeline

What are the key characteristics that make something a `transform`? I would say it's a function that takes a `relation` as the last input parameter and produces a `relation` as output.

What makes something a `pipeline`? It's a function that produces a `relation` as output but which doesn't take a `relation` as the last input parameter. I think it is important to distinguish this from the `relation` type itself which is purely a data declaration type which you might want for external data declarations as in the discussion in #381 (and rslabbert's comment in particular).

**Reply by snth** [reply]
Maybe I should use `() -> relation` as the type signature for `pipeline` rather than `null -> relation`. I just mean something that takes no inputs/parameters, e.g. like the following signature for `random` in Rust `pub fn random<T>() -> T`.

My main point there is that while the return type is `T`, the signature is distinct from `T` itself.

**Reply by snth** [reply]
I guess another term for `pipeline` would be `query`. A `query` when applied in a database produces a `relation` but they are distinct. In particular the same `query` applied in different databases could produce different `relations`.

**Reply by aljazerzen** [reply]
Oh, I was confused about `null -> relation`, that makes sense.

The distinction between our interpretations is in this:

>  while the return type is T, the signature is distinct from T itself.

I'm treating `() -> T` to be equivalent to `T`. I'm not sure why or when I started thinking about PRQL function this way, but this is a common practice in functional languages.

While it does need an assumption that all your functions are [pure](https://en.wikipedia.org/wiki/Pure_function), it simplifies dealing with function calls, currying and pipelines.

In PRQL we don't have syntax for a function call without arguments, so it's hard to pin-point when the call should happen.

**Reply by aljazerzen** [reply]
If we want to make the distinction between `() -> T` and `T`, we should start writing the queries in c-style syntax or invent syntax for evaluation of function without args.

**Reply by aljazerzen** [reply]
Currently, compiler (and thus the language) does not make this distinction. Is there a reason that we should change that?

**Reply by snth** [reply]
Not necessarily, in my mind I just have them sitting in slightly different categories.

The way I think about the `relation`/`table` is as a concrete table definition in a database. As per rslabbert's comment in #381 I pictured us connecting to a database, downloading the schema definitions and storing these in *"declaration files"*, I guess kinda like `DDL`, perhaps both `table` and `view` being subtypes of `relation`. You could store such declaration files in different directories for different databases, say *prod*, *uat* and *dev*. A `pipeline` would be a PRQL `query` which runs against one of these directories/environments and then evaluates to a resultant `relation` type, perhaps writing out a *declaration file* of it's own. A `pipeline` with signature `() -> T` on the other hand would be `DQL`. I don't know if the distinction is important from the compiler's perspective, to me they just seem like separate concepts.

So `pipelines` and `transforms` live in the PRQL Query space/category and then the PRQL Compiler is a Functor that transforms them to SQL in the Database space/category. `relations` are like table definitions that define the shape of data, and which "live" at the interface of consecutive transforms.

Through type inference you could hopefully figure out what schema constraints your source ` relations` have to satisfy for a particular `transform` or `pipeline` so that you could easily check this against a particular `relation` declaration file to figure out if it's compatible or not.

**Reply by aljazerzen** [reply]
Well, I see most of the things you described exactly the same. With the extension that PRQL `query` (RQ) is also a `relation` - just a computed one.

**Reply by max-sixty** [reply]
I think this is exactly how I think about it.

I'm not sure exactly where Pipeline fits here — it seems that Relation and Transform is fairly complete. But we are somewhat bound to the Pipeline term given the PRQL name! And it does intuitively fill a useful space where "Relation followed by a series of Transforms" isn't welcome. 

So maybe a Pipeline means a series of transforms, possibly bound to a relation?

**Reply by snth** [reply]
> So maybe a Pipeline means a series of transforms, possibly bound to a relation?

That's how I think about the term pipeline and how I have been using it above. However in the book currently pipeline seems to mean (possibly multi-step) transform (eg as an argument to [`group`](https://prql-lang.org/book/transforms/group.html)) while `table` is used to refer to relations (eg as an argument to [`join`](https://prql-lang.org/book/transforms/join.html), [`from`](https://prql-lang.org/book/transforms/from.html), and [`concat` and `union`](https://prql-lang.org/book/transforms/concat-and-union.html)).

I'm happy to adopt the terminology of the book but I just think we need to pick one convention, spell it out explicitly somewhere and then use it consistently because the concepts are not interchangeable, i.e. `join` and `group` take arguments of different types and we should have names for them.

Given that the P in PRQL stands for pipeline, I think the `pipeline` term should be given a prominent role though.

**Reply by aljazerzen** [reply]
I think of pipeline more like just a syntactic construct - like binary operation or function call. I'd also call this a pipeline:
```
(12 | div 5 | floor | mul 5)
```
... it's just that it doesn't evaluate to a relation, but to a number.

So:
- group takes a function (usually expressed as a pipeline),
- from and join take a relation (possibly expressed as a pipeline),
- concat takes a relation (possibly expressed as a pipeline).

I'd say that the name of the language implies that it uses this construct as a core feature.

**Comment by aljazerzen** [comment]
Regardless of the type of a variable (be it table, relation, function, transform or pipeline), I think we can change syntax of variable declaration to use let and an optional type annotation:
```diff
-table a = (...)
-pipeline a = (...)
+let a <relation> = (...)
```

**Reply by snth** [reply]
And what about function definitions?

**Reply by aljazerzen** [reply]
```
func my_func a b -> (...)
```
... can stay as a syntactic sugar for:
```
let my_func = a b -> (...)
```
... or something similar, we can still decide on syntax for lambda functions (and if we even need them).

**Reply by snth** [reply]
I guess I don't really understand what advantage it confers and I worry that the term `let` will become quite overloaded.

At the moment we have this nice property that you can mostly scan down the left hand column and get a good idea of what happens on that line, e.g.

```elm
func a ...
func b ...
table c ...

from d
filter ...
```

With `let` you would have to look at the type annotation or the value definition to figure out what the constant/value is.

It might be a worthwhile price to pay but I don't understand the benefit yet.

**Reply by aljazerzen** [reply]
Having `let` also tells you what's happening in this line: a new thing gets defined and can later be used.

C-style languages, did use the type name as start of var def, but modern languages are switching to have let / const / var / auto, because of things like this:
```
String a = ...;
MyClassDelagator b = ...;
int c = ...;
```
as opposed to:
```
let a: String = ...;
let MyClassDelagator = ...;
let c: int = ...;
```

**Reply by aljazerzen** [reply]
And I don't know about anything else using `let` ATM.

**Comment by aljazerzen** [comment]
Regarding custom transforms: this is implemented but not documented:
```
func take_top_salaries n rel -> (
  # first expr in the pipeline is the value
  # in our case, the input relation 
  rel
  # then any transforms we want
  # note that we can refer to column in the relation
  # ... if they exist of course
  sort [-s.salary]
  take n
)

# this will work
from s=salaries
take_top_salaries 5

# but this will not: no relation s
from salaries
take_top_salaries 5
```

**Reply by aljazerzen** [reply]
Something we'd need to explain is this:
```
1: from albums
2: take 3

   # column "title" is not lookup up in the relation "albums" defined on line 1
   # it is looked up in the second argument of select function   
3: select title

# it would be resolved as:
(take 3 (from albums)).title

# even if we used "albums.title":
(take 3 (from albums)).albums.title
```

**Reply by snth** [reply]
I initially didn't understand your second example but after rereading it I think I'm starting to catch on but please bear with me. I'll try and rephrase it in terms of the first example to check if I understand:

What's possibly surprising in your `take_top_salaries` definition (and where I think your second example was going), is that `rel` doesn't appear anywhere in the definition. So would it be possible to change it to:
```elm
func take_top_salaries n rel -> (
  sort [-rel.salary]
  take n
)
```
so that
```elm
from s=salaries
take_top_salaries 5
```
is interpreted as
```elm
from s=salaries
sort [-s.salary]
take n
```
and
```elm
from salaries
take_top_salaries 5
```
as
```elm
from salaries
sort [-salaries.salary]
take n
```
? Or in terms of your second example
> it is looked up in the second argument of select function

Let's call it `rel` as well, i.e. `rel == (take 3 (from albums))` in that case?

**Reply by snth** [reply]
Or another example
```elm
from equity
derive [salary = stock_options]
take_top_salaries 5
```
Could that work? Could we make it work? Would we want to make it work?

I guess the discussion I'm trying to foster is how do we structure PRQL so that we can make the most of the composability so that we can reap the same benefits of functional composability and reusable code as in other languages?

**Reply by aljazerzen** [reply]
>  So would it be possible to change it to: `sort [-rel.salary]`

No, this wouldn't be possible.

I'm trying to say that `rel.salary` or `s.salary` is looked up - as is, using `rel.salary` as the column name - in the second argument of `sort`. In case of:

```elm
func take_top_salaries n rel -> (
  sort [-rel.salary]
  take n
)
# or equivalently: 
func take_top_salaries n rel -> (
  take n (sort [-rel.salary])
)
```
... there is no second argument! So this would error out, saying that `sort` is missing an arg.

But if we do like this:
```elm
func take_top_salaries n rel -> (
  take n (sort [-rel.salary] rel)
)
```
... we'd be looking up `rel.salary` in `rel`. In cases you provided, no relation had a column `rel.salary`, so it would be also produce an error.

**Reply by snth** [reply]
Btw, the fact that the top example works at all is amazing! 👏 I think it's an incredible feature for PRQL to have and something we should definitely document and highlight as it really sets it apart.

I don't want my somewhat theoretical discussions of what could be possible in theory (or in an ideal language) to detract from that in any way. I've just been trying to figure out for myself to what extent one could start writing reusable functions and packages in PRQL and where the barriers might be.

**Reply by aljazerzen** [reply]
I enjoy this theoretical discussion a lot :smile:  It brings up corner cases that probably only I know of. Which means that their design may be very biased and in need of a second opinion.

So do try to compose more examples so we can discuss if they make sense in my interpretation and if the compiler should be adjusted.

**Reply by snth** [reply]
Just linking to #1335 as there is a lot of discussion about namespaces there.
